crates/amaru-ledger/src/rules/transaction/collateral.rs

```diff
@@ -0,0 +1,312 @@
+use std::collections::BTreeMap;
+
+use amaru_kernel::{
+    protocol_parameters::ProtocolParameters, AlonzoValue, MemoizedTransactionOutput,
+    MintedTransactionOutput, TransactionInput, TransactionInputAdapter, Value,
+};
+use thiserror::Error;
+
+use crate::context::UtxoSlice;
+
+struct CollateralBalance {
+    pub coin: u64,
+    pub multiasset: BTreeMap<Vec<u8>, u64>,
+}
+
+impl From<&Value> for CollateralBalance {
+    fn from(value: &Value) -> Self {
+        match value {
+            Value::Multiasset(coin, multiasset) => {
+                let mut map = BTreeMap::new();
+                multiasset.iter().for_each(|(policy, assets)| {
```

Comment:
Also smells like could be do without `for_each`

Reply:
Sure, could be a map, save a few lines!